include/picongpu/particles/densityProfiles/IProfile.hpp

```diff
@@ -40,14 +40,41 @@ namespace picongpu
         template<typename T_Profile>
         struct IProfile : private T_Profile
         {
-            /** Create a profile functor for the given time iteration
+            /** create a profile functor for the given time iteration
              *
-             * @param currentStep current time iteration
+             * This constructor is only compiled if the user functor has
+             * a host side constructor with one (uint32_t) or (uint32_t, IdGenerator) arguments.
+             *
+             * @tparam DeferFunctor is used to defer the functor type evaluation to enable/disable
+             *                      the constructor
+             * @param currentStep current simulation time step
+             *
+             * @{
              */
-            HINLINE IProfile(uint32_t const currentStep) : T_Profile(currentStep)
+            template<typename DeferFunctor = T_Profile>
+            HINLINE IProfile(
+                uint32_t currentStep,
+                IdGenerator&,
```

Comment:
Small comment, not blocking.
IdGenerator is passed sometimes by reference and sometimes by value. It only holds a pointer, so it isnt a big deal, but it can be made more uniform. Also the argument can be marked as `const`.
This happens in many places in this PR